### PR TITLE
go/store/datas/pull: pull_table_file_writer.go: Optimize pull/push to allow for concurrent table file uploads.

### PR DESCRIPTION
**Pull request by reltuk**
Move management of the upload workers and the table file writer to its own struct which is testable in isolation.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>edbed08</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>edbed08</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>9e99878</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>9e99878</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3fa1a00</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3fa1a00</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>10bfb3c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>10bfb3c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5cbcb7b</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5cbcb7b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>9fe73a3</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>9fe73a3</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/datas/pull/pull_table_file_writer.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+			// Cleanup dangling writer, whose contents will never be used.
+			curWr.Finish()
+			rd, _ := curWr.Reader()
+			if rd != nil {
+				rd.Close()
+			}
+		}
+	}()
+
+	for {
```
Comment (macneale4) [inline go/store/datas/pull/pull_table_file_writer.go:170]: The bulk of review time was trying to understand the state machine of this loop. Having channels going nil then coming back in certain situations make this for a pretty hard loop to follow.

I don't have anything to suggest other than add more comments to spare the next reader from needing to figure this out. Maybe break the `newChnk, ok := <-addChunkCh:` case into a seperate function? Having the select block be really clear about the different assumptions when you indifferent states may help.

**Review by macneale4 — CHANGES_REQUESTED** [review]

File: go/store/datas/pull/pull_table_file_writer.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+		select {
+		case <-ctx.Done():
+			return eg.Wait()
+		case <-w.egCtx.Done():
+			return context.Cause(w.egCtx)
+		case thisReqCh <- pendingUpload:
+			// Keep track of how many responses we need to wait for.
+			outstandingUploads += 1
+			pendingUploads[len(pendingUploads)-1] = uploadReq{}
+			pendingUploads = pendingUploads[:len(pendingUploads)-1]
```
[the remark this answers is not in the document]
Comment (macneale4) [reply go/store/datas/pull/pull_table_file_writer.go:200]: I realized offline that you are setting to the empty uploadReq to free up the table writer. Probably worth a comment.

**Review by macneale4 — APPROVED** [review]
LGTM

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>43bc4cd</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>43bc4cd</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**